qutebrowser/javascript/webelem.js

```diff
@@ -184,14 +225,67 @@ window._qutebrowser.webelem = (function() {
             }
         }
 
+        // Recurse into frames and add them
+        for (let i = 0; i < subelem_frames.length; i++) {
+            if (iframe_same_domain(subelem_frames[i])) {
+                const frame = subelem_frames[i];
+                const subelems = frame.document.
+                    querySelectorAll(selector);
+                for (let elem_num = 0; elem_num < subelems.length; ++elem_num) {
+                    if (!only_visible ||
+                        is_visible(subelems[elem_num], frame)) {
+                        out.push(serialize_elem(subelems[elem_num], frame));
+                    }
+                }
+            }
+        }
+
         return out;
     };
 
+    // Runs a function in a frame until the result is not null, then return
+    function run_frames(func) {
+        for (let i = 0; i < window.frames.length; ++i) {
+            const frame = window.frames[i];
+            if (iframe_same_domain(frame)) {
+                const result = func(frame);
+                if (result) {
+                    return result;
+                }
+            }
+        }
+        return null;
+    }
+
     funcs.find_id = (id) => {
         const elem = document.getElementById(id);
-        return serialize_elem(elem);
+        if (elem) {
+            return serialize_elem(elem);
+        }
+
+        const serialized_elem = run_frames((frame) => {
+            const element = frame.window.document.getElementById(id);
+            return serialize_elem(element, frame);
+        });
+
+        if (serialized_elem) {
+            return serialized_elem;
+        }
+
+        return null;
     };
 
+    // Check if elem is an iframe, and if so, return the result of func on it.
+    // If no iframes match, return null
+    function call_if_frame(elem, func) {
+        const frame = elem.contentWindow;
+        // Check if elem is a frame, and if so, call func on the window
+        if (frame && iframe_same_domain(frame) && frame.frameElement) {
```

Comment:
Hmm, are you sure this works correctly? Looking at the docs for [contentWindow](https://developer.mozilla.org/en-US/docs/Web/API/HTMLIFrameElement/contentWindow) and [frameElement](https://developer.mozilla.org/en-US/docs/Web/API/Window/frameElement), it seems to me like `contentWindow` is only defined on frames (so this would probably crash?), but `frameElement` gets you the frame of *any* element which is inside a frame.

Reply:
I'm fairly certain this works as is, since this is used for insert mode on click, and that seems to be working right now for me.

In Javascript (I think), if you dereference something that dosen't exist, you'll get undefined, so document.nonexistent will return undefined. In this line, I check to see if frame is truthy, and then if `frame.frameElement` is truthy, and in that way, I check to see if `elem.contentWindow.frameElement` exists.

I asked around, and `'contentWindow' in elem` is an alternative to this, would that be better to use?

From what I tested, every element in a frame dosen't seem to have a `frameElement` property, only the top window does (and that's what the docs I could find on it say as well). https://developer.mozilla.org/en-US/docs/Web/API/Window/frameElement

I could be missing something though, so let me know if that's the case :)

Reply:
Hmm, right. I thought `"use strict";` would prevent the `undefined` behaviour (hah 😆), but that doesn't seem to be the case. I think something like `'contentWindow' in elem && iframe_same_domain(elem.contentWindow) && 'frameElement' in elem` would definitely be more readable, even if a bit longer.